app/models/classification_pipeline.rb

```diff
@@ -90,13 +90,24 @@ def reduce(reducible_id, subject_id, user_id, extract_ids=[])
     return [] unless reducers&.present?
     retries ||= 2
 
+    prior_extracts = []
+    if subject_id
+      subject = Subject.find(subject_id)
+      prior_subject_ids = subject.additional_subject_ids_for_reduction
+      prior_extracts = Extract.where(subject_id: prior_subject_ids).pluck(:id)
```

Comment:
Okay, so what about this: instead of going to load the prior extracts, and then plucking their IDs so that the extract fetcher can then load them, we just add the additional subject IDs into the filter clause. So instead of this line, do:

```suggestion
      subject_id = [subject_id] + prior_subject_ids
```

Then the fetcher doesn't need to get any knowledge about "loading additional subjects", because it'll just execute the filter and Rails will automatically do a `SELECT ... WHERE subject_id IN (1,2,3)` rather than `WHERE subject_id=1`.

Reply:
The whole pipeline is kind of hard to follow, tbh. I think I know what you mean, but from here it seems to be doing the same thing in both strategies. What is `exact_extracts` _supposed_ to mean if its not to pull the exact extracts passed to the fetcher? This is the functionality I wanted to expand: to be able to inject a specific set of ids in addition to the ones found by the subject scope. And that's why I gave it a new strategy, so that it would do both things in a way that could be explicitly set. 

I don't think it's less confusing to override the subject_id and make it an array that contains more than a single subject id, but keeping the fetcher entirely out of the equation does keep it contained to the classification pipeline. If you feel strongly about this I'll get in a quick commit to change it.

Reply:
I'm with Zach here, I think. This is what `exact_extracts` was supposed to mean. It's just a little strange to take the exact ids we told the fetcher to get and then expand that list, but I see the utility. Maybe the only thing I'm asking for is to `#uniq` the list of ids before fetching them.

As for the pipeline code, I agree that it's getting quite messy. At some point in the past, @marten suggested blowing up the whole class and moving the code it contains to the relevant classes.

Reply:
Okay, but with this new mode it's not really **exact** extract anymore, since it's loading more than exactly those extracts.

Reply:
True enough, but that word isn't used anywhere outside of the extract fetcher, so changing it there wouldn't be terribly impactful if that's your main issue. I like either `explicit_extracts` or `defined_extracts`. 

That said, it would also make sense to me to make the extant strategies include any passed-in ids by default, with `fetch_minimal` to bypass the scopes. I didn't love making the 'additional' strategy subject-only, but I didn't want to add two new ones. That might be a clearer way to go.

And @amyrebecca I'll replace the concat with a union op and that should prevent any dupe ids from making it into the fetcher, thanks.

Reply:
I think the second paragraph makes a lot of sense. Something about having another strategy just for this specific combination rather than `defined_extracts` just being a no-op when empty seemed weird, good shout on that.